### PR TITLE
doc: add nested virtualization user guide

### PR DESCRIPTION
**Pull request by ZideChen0**
Tracked-On: #5923
Signed-off-by: Zide Chen <zide.chen@intel.com>

### CODE REVIEW

**Comment by ZideChen0** [comment]
Accidentally I deleted my branch which automatically deleted the previous PR #6083

Diff to #6083:

1. addressed all the review comments
2. added a new section "Create an Ubuntu KVM Image"
3. used screen shots from acrn-config GUI to show the ACRN configurations
4. explicitly mention that pre-launched is not officially supported to run nested.
5. some wording, especially in first paragraph.

File: doc/tutorials/nvmx_virtualization.rst
Comment (dbkinder) [inline doc/tutorials/nvmx_virtualization.rst:43]: Would you add an explanation of the ``VMCS01``, ``VMCS02``, and ``VMCS12`` names meanings.  I presume all VMs have a VM control structure?  Also, the figure below has nothing labeled as the "L0 hypervisor" or "L2 hypervisor" mentioned in the text in the figure, Should the L2 hypervisor be L1 hypervisor (or maybe L2 VM)?

The text bullets in the image should be pulled out and added instead as bullets in the text so it could be easily edited and searched.

File: doc/tutorials/nvmx_virtualization.rst
Comment (dbkinder) [inline doc/tutorials/nvmx_virtualization.rst:114]: If this is a limitation of the configuration editor, we should fix the editor!  @junjiemao1 @NanlinXie

**Review by dbkinder — CHANGES_REQUESTED** [review]
@ZideChen0 I went through and made some clarity and formatting edits an force pushed my changes.  I still have some questions though that need your attention.

File: doc/tutorials/nvmx_virtualization.rst
Comment (dbkinder) [inline doc/tutorials/nvmx_virtualization.rst:267]: @NanlinXie @junjiemao1 Can the ``vcpu_list`` and ``vm_list`` commands be updated to show these nested VMs?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:267]: > @NanlinXie @junjiemao1 Can the `vcpu_list` and `vm_list` commands be updated to show these nested VMs?

No, nested VMs are managed by L1 hypervisor not ACRN, so  HV commands can't show them.

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:114]: > If this is a limitation of the configuration editor, we should fix the editor! @junjiemao1 @NanlinXie

Without nVMX feature, the CPU affinity assignment of SOS is considered as "all pCPUs minus pCPUs assigned to pre-launched VMs" so it's no need to edit CPU assignment for SOS. Since nested is considered as an experimental feature at this moment and the VM configuration requirement could be changed in the future, I guess we don't have to do this change for now.

Comment (dbkinder) [reply doc/tutorials/nvmx_virtualization.rst:114]: OK.  In that case, you can delete this section.

File: doc/tutorials/nvmx_virtualization.rst
Comment (dbkinder) [reply doc/tutorials/nvmx_virtualization.rst:267]: Is there an L1 hypervisor (KVM) command to show these nested VMs?

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:43]: There is no "L2 hypervisor", will fix the typo and add "L0 hypervisor" to the figure (:name: nested_virt_hld). The concept of VMCS01, VMCS02, and VMCS12 are not very straightforward, and I can add more explanation in the HLD doc.

Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:43]: > Would you add an explanation of the `VMCS01`, `VMCS02`, and `VMCS12` names meanings. I presume all VMs have a VM control structure? Also, the figure below has nothing labeled as the "L0 hypervisor" or "L2 hypervisor" mentioned in the text in the figure, Should the L2 hypervisor be L1 hypervisor (or maybe L2 VM)?
> 
> The text bullets in the image should be pulled out and added instead as bullets in the .rst file so it could be easily edited and searched.

Yes, good idea!

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:131]: > OK. In that case, you can delete this section.

But with nVMX enabled, the user may need to edit the CPU affinity for SOS.

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:284]: AFAIK there is no such command available in Linux to show QEMU launched guests.

**Review by dbkinder — APPROVED** [review]
LGTM, thanks!  If @gvancuts has no issues, we're good to merge and publish.

**Comment by ZideChen0** [comment]
> LGTM, thanks! If @gvancuts has no issues, we're good to merge and publish.

@dbkinder  before it's merged, I need to do another push --force to update the test platforms. (no changes to anything else).

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [reply doc/tutorials/nvmx_virtualization.rst:284]: I'm not aware of any other mean of listing these QEMU VMs short of grepping for all `qemu*` processes running on the host.

What we could do as an enhancement to this guide is to create the VM and start/stop it using `virsh`. That way you can also list them all.

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:76]:
```suggestion
on Tiger Lake and Kaby Lake platforms (See :ref:`hardware`.)
```

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:88]:
```suggestion
* Only the ``SCHED_NOOP`` scheduler is supported. ACRN can't receive timer interrupts
```

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:210]: This alone will not install the kernel and modules. We can refer to the GSG section here for that, or repeat them (i.e. copy `bzImage` and `sudo make modules_install`)

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:195]: A bit of a side question: could we turn these on already today by default or are there side-effect to having those?

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:256]: Does that automatically attach the VM to the `acrn-br0` bridge?

Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:256]: No acrn-br0 is needed her. User can check the generic QEMU documentation on how to launch nested guests.

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:195]: kernel_config_uefi_sos is for SOS which doesn't need KVM, not really any side-effect. Since nested is considered as an experimental feature, I think it's good to keep kernel_config_uefi_sos  clean.

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:210]: In the beginning of this section, I refer to "rt_industry_ubuntu_setup" on to prepare kernel and rootfs.

File: doc/tutorials/nvmx_virtualization.rst
Comment (ZideChen0) [reply doc/tutorials/nvmx_virtualization.rst:284]: Yes, that'a true. I think this guide is focusing on ACRN and the 'nested' specific features.In terms of how to launch guest from SOS, it's not important, and users can learn it from generic QEMU/libvirt documentations.

**Comment by dbkinder** [comment]
@wenlingz @NanlinXie This doc-only change shouldn't be failing a code build check :(

**Comment by dbkinder** [comment]
OK to check

**Comment by wenlingz** [comment]
OK to verify

**Comment by wenlingz** [comment]
> @wenlingz @NanlinXie This doc-only change shouldn't be failing a code build check :(

There had network issue , didn't run the scripts, so failed.

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [reply doc/tutorials/nvmx_virtualization.rst:195]: OK, let's keep it clean for now. We can revisit this later.

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:199]: You can now remove this - see https://github.com/projectacrn/acrn-kernel/pull/338

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [reply doc/tutorials/nvmx_virtualization.rst:210]: You're right. Let's just remove this `code-block` altogether then?

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [reply doc/tutorials/nvmx_virtualization.rst:256]: I understand it's not strictly needed, but what happens if it is present (because of other User VMs being run on the system for example). Are we re-using it? Can ACRN-launched User VMs communicate with QEMU Guest VMs if both are running on the same system?

File: doc/tutorials/nvmx_virtualization.rst
Comment (gvancuts) [inline doc/tutorials/nvmx_virtualization.rst:186]: I would remove this line. This is what's checked out by default already.

**Pull request merged**
